### DISCUSSION TITLE
is webGPU actually web-specific somehow?

### DISCUSSION THREAD

**Discussion by LastExceed** [gfx-rs/wgpu#4251] [Q&A]
I'm at the very start of learning how to interact with a GPU, so i currently know next to nothing. While searching for a starting point i came across this crate. the name "webGPU" and its history strongly suggest that this is for web applications, but looking into the documentation it seems like web is nothing more than another possible platform to target, and working with wgpu doesn't necessarily involve any webdev at all. I wondering, is this some web thing ported to desktop (with corresponding overhead, like electron disguising a browser as a native desktop app), or is it platform native by design with just a misleading name? would it be a misuse to choose this crate when not targeting web at all?

sorry if my question is lacking a bit of structure, asking clear isolated questions is difficult with this little knowledge

**Comment by Ciubix8513** [comment]
Well wgpu is a sort of small abstraction layer on top of native graphics api, which include WebGPU, so no it's not just a web thing or like electron. But i think it does has a small overhead compared to interacting with a native api, but that overhead is practically insignificant. And no it's not a misuse of the crate to use outside of web, Veloren uses wgpu, Bevy uses wgpu, Iced uses wgpu, and a lot of other projects

**Reply by LastExceed** [reply]
so `wgpu` is just an abstraction over lots of graphics APIs (webGPU being one of them) and not a wrapper for webGPU specifically?

**Reply by Ciubix8513** [reply]
Yes

**Reply by LastExceed** [reply]
then can you explain what this sentence from the readme means?
> The api is based on the [WebGPU standard](https://gpuweb.github.io/gpuweb/). It serves as the core of the WebGPU integration in Firefox, Servo, and Deno.

**Reply by Ciubix8513** [reply]
well, it's just that you use wgpu by writing code the way you would with WebGPU, and it then translated into other graphics apis

**Reply by LastExceed** [reply]
oh so it just used webGPU as a design guideline and is otherwise not specific to it any more than other APIs ?

**Reply by Ciubix8513** [reply]
Yea

**Comment by kpreid** [comment]
WebGPU has these requirements in order to serve its goals:

* It must be impossible to use the API such that it causes memory corruption or can read data the application did not have access to — it must be memory safe.
* it must be implementable on top of common GPU driver APIs.

Conveniently, both of these things *also* make the WebGPU API design a good API design for a Rust GPU abstraction layer. Thus, `wgpu` is a useful Rust library.

And, in addition to its use in Rust applications, parts of `wgpu` are used to _implement_ WebGPU in Firefox, since `wgpu` has got the abstraction layer even if it doesn't have a JavaScript front end — you don't need to care about this, but it explains why `wgpu` is not just "inspired by" WebGPU but takes care to be conformant to the standard itself. (So if you write a Rust application with `wgpu`, and compile it to WebAssembly and run it in Firefox, you could be running `wgpu` on top of `wgpu`!)
